### PR TITLE
Autoclose blocks when typing `{`

### PR DESCRIPTION
**Pull request by jonas-schievink**

### CODE REVIEW

**Comment by jonas-schievink** [comment]
Hmm, it's hard to make this trigger in all conditions I want it in, but not in this:

```rust
fn f() {
    $0assert!(true);
    let a = 0;
}
```

...because it's unclear how many statements are intended to be included in the block. This would probably be easier if we could parse the file *before* the change that inserted `{`, is there any way to do that?

**Review by SomeoneToIgnore — COMMENTED** [review]
I wanted that for quite a while, thank you! 🚀 

Uplifting to see it working for closures, but with match arms something interesting happens:
```rust
match Some(Some(2)) {
    Some(Some(a)) => something(a),
    Some(None) => 22,
    None => 33,
}
```

here it works for `None =>` variant, but not for the others.


I would imagine it could work in various let bindings like
`type_char('{', r"let _a = $0vec![1];", r"let _a = {vec![1]};");`
but not sure if feasible to implement now.

**Comment by jonas-schievink** [comment]
I've pushed my local version which triggers more often, but it also triggers in undesirable situations like the one in https://github.com/rust-analyzer/rust-analyzer/pull/8388#issuecomment-814491201.

**Comment by jonas-schievink** [comment]
Oh, this seems super broken when typing `{` in a comment, that needs a fix

**Comment by matklad** [comment]
> This would probably be easier if we could parse the file before the change that inserted {, is there any way to do that?

Yes, see how we insert fake ident during completion. Basically, just remove { from text and re-parse. Sadly, our incremental parsing won’t work at the moment for this use-case, but it is not a fundamental limitation.

File: crates/ide/src/typing.rs [hunk 1]
```diff
@@ -70,15 +71,39 @@ fn on_char_typed_inner(file: &SourceFile, offset: TextSize, char_typed: char) ->
...
         '>' => on_arrow_typed(file, offset),
+        '{' => on_opening_brace_typed(file, offset),
         _ => unreachable!(),
     }
 }
 
+/// Inserts a closing `}` when the user types an opening `{`, wrapping an existing expression in a
+/// block.
+fn on_opening_brace_typed(file: &SourceFile, offset: TextSize) -> Option<TextEdit> {
+    stdx::always!(file.syntax().text().char_at(offset) == Some('{'));
```
Comment (matklad) [inline crates/ide/src/typing.rs:82]:
```suggestion
    if !stdx::always!(file.syntax().text().char_at(offset) == Some('{')) { return None; }
```

File: crates/ide/src/typing.rs [hunk 2]
```diff
@@ -373,4 +398,39 @@ fn main() {
...
     }
+
+    #[test]
+    fn adds_closing_brace() {
+        type_char('{', r"fn f() { match () { _ => $0() } }", r"fn f() { match () { _ => {()} } }");
+        type_char('{', r"fn f() { $0(); }", r"fn f() { {()}; }");
+        type_char('{', r"fn f() { let x = $0(); }", r"fn f() { let x = {()}; }");
+        type_char(
+            '{',
+            r"
```
Comment (matklad) [inline crates/ide/src/typing.rs:409]: Let’s use standard  test style, with hashes and without indentation:

https://github.com/rust-analyzer/rust-analyzer/blob/master/docs/dev/style.md#minimal-tests

**Comment by jonas-schievink** [comment]
> Oh, this seems super broken when typing `{` in a comment, that needs a fix

Resolved

**Review by SomeoneToIgnore — APPROVED** [review]
Not sure if worth it, but there are a few negative cases that might be improved:

`let zz = $0'a'};` now turns into `let zz = {'a'}};`

Yet not sure those are frequent to be worth fixing.

**Comment by jonas-schievink** [comment]
Ah, good example. That seems pretty hard to fix though, at least without looking at the encountered parse errors.

**Comment by jonas-schievink** [comment]
Been using this for a while and it hasn't gotten in the way, so I'd say let's land this

bors r+

**Comment by matklad** [comment]
I've just recalled something! One of the original design goals of rust-analyzer was to have every user-visible feature gated by a flag. Some features are very annoying to some users, so it is valuable to be able to disable anything that you don't need. 

We didn't historically fulfill this vision, as our configuration was a mess (it was a global, mutable, stringly-typed store of data). Now that we have a single source of truth for config in `config.rs`, and that we figured a simple pattern of just passing in specific, narrow `Config` structs to various methods, I wonder if we should go back to that original vision. 

Specifically, do we want to introduce `rust.analyzer.typing.(indentDot|addSemi|wrapInBraces)` configs here, as we are adding a new feature anyway?

If we do, we should also add `Configuration` section as a cross-cutting concern into Architecture.md.

**Comment by SomeoneToIgnore** [comment]
This particular feature is behind the `editor.formatOnType` flag, so goal nominally fulfilled? 🙂 

My biased perception sees this feature as the less potentially annoying, compared to the other two similar completions, since it's harder to trigger.
The code also looks relatively safe performance-wise to leave it be and wait for the explicit user feedback for a feature flag.

If not feedback comes, we can wait and utilize the idea with the next similar completion added: I personally have an idea to similarly complete various `<_, <_, _>>` mazes.

**Comment by matklad** [comment]
> This particular feature is behind the editor.formatOnType flag, so goal nominally fulfilled? 🙂


not really, the idea was to have reeeeealy granular features (a feature for each separate assist/fixit/diagnostic/etc)

**Comment by lnicola** [comment]
changelog feature

**Comment by jonas-schievink** [comment]
gif for changelog:

r-a braces

**Comment by matklad** [comment]
Looking at the gif, we should change our OnEnter to snap that } to the other line.

**Comment by SomeoneToIgnore** [comment]
Somewhat related: if I want to revert the result, it takes two "undo", one for each curly, which is somewhat counter-intuitive compared to a single action I needed to trigger this completion initially.

Not sure if that's anyhow fixable for all editors though.

**Comment by Milo123459** [comment]
> Looking at the gif, we should change our OnEnter to snap that } to the other line.

This would make it considerably better in my opinion

**Comment by jonas-schievink** [comment]
Yeah that sounds like a great idea, I'll play around with it

**Comment by flodiebold** [comment]
Does that need to be in OnEnter, or can it be done in OnTypeFormatting?

**Comment by matklad** [comment]
Better to double check, but when I looked into this couple of years ago, OnEnter was required.

**Pull request merged**